scalafmt: add code formatting, improve coding-style compliance

kitbellew opened the pull request:
Based on #4522.

kitbellew commented on .scalafmt.conf:
```diff
@@ -0,0 +1,130 @@
+# Test upgrades: $ scripts/scalafmt --test 2> diff.txt
+version = "3.10.3+10-2b4fdf2e-SNAPSHOT"
```
this is the most recent commit. if adjustments are requested and implemented, this will change. when this PR is about to be accepted, I will release and update to a released version.

kitbellew commented:
@sjrd the job that failed uses scala-2.12.

if 2.12 is still used, then the dialect specified in the configuration file is incorrect (scala213source3). should it be replaced with scala212 or scala212source3 (if you use -Xsource:3)?

sjrd commented:
We do not use `-Xsource:3`.

ekrich commented on compiler/src/main/scala/org/scalajs/nscplugin/ExplicitLocalJS.scala:
```diff
@@ -212,7 +212,8 @@ abstract class ExplicitLocalJS[G <: Global with Singleton](val global: G)
             decl match {
               case ClassDef(_, _, _, impl)
-                  if decl.symbol.isModuleClass && isInnerJSClassOrObject(decl.symbol) =>
+                  if decl.symbol.isModuleClass && isInnerJSClassOrObject(
```
I think this is a manual change to but wrapping after the `&&` is nicer to keep the function `isInner...` together. In general this is what is nice about scalafmt; if the copy wraps and it is not desired, you can usually adjust the wrap and get something nicer.

kitbellew commented on ir/shared/src/main/scala/org/scalajs/ir/UTF8String.scala:
```diff
@@ -174,7 +175,8 @@ object UTF8String {
         val b4 = bytes(i + 3) & 0xff
-        if (isInvalidNextByte(b2) || isInvalidNextByte(b3) || isInvalidNextByte(b4)) {
+        if (isInvalidNextByte(b2) || isInvalidNextByte(b3) || isInvalidNextByte(
+                b4)) {
```
another option is to use a parameter which breaks after some infix operators automatically. but then it might not allow you to choose where :) and that will add more diffs.

kitbellew commented on examples/reversi/src/main/scala/reversi/Reversi.scala:
```diff
@@ -82,15 +87,16 @@ class Reversi(jQuery: JQueryStatic, playground: JQuery) {
-    val HalfSquareSizePx = SquareSizePx/2
-    val PawnRadiusPx = HalfSquareSizePx-4
-    val BoardSizePx = BoardSize*SquareSizePx + 3
+    val HalfSquareSizePx = SquareSizePx / 2
```
well...
- there's a parameter
  - https://scalameta.org/scalafmt/docs/configuration.html#spacesaroundsymbolicinfixoperators
- but coding style says "Binary operators must have a single space on both sides" 
  - https://github.com/scala-js/scala-js/blob/main/CODINGSTYLE.md#spaces

kitbellew commented on compiler/src/main/scala/org/scalajs/nscplugin/PreTyperComponent.scala:
```diff
@@ -116,7 +117,7 @@ abstract class PreTyperComponent(val global: Global)
       case vdef: ValDef => needsAnnotations(vdef)
       case ddef: DefDef => needsAnnotations(ddef)
-      case _ => false
+      case _            => false
```
i don't use alignment... but others do: https://github.com/scala-js/scala-js/blob/main/CODINGSTYLE.md#pattern-matching

ekrich commented:
ok, it sounds like most of these things follow the coding standard so that would have to change to keep those changes provided scalafmt supports it.

ekrich commented on javalib/src/main/scala/java/lang/Class.scala:
```diff
@@ -21,8 +21,8 @@ final class Class[A] private ()
-        if (isPrimitive()) "" else "class ")+getName()
+    (if (isInterface()) "interface "
+     else if (isPrimitive()) "" else "class ") + getName()
   }
```
I guess that leaves `val prefix = ... prefix + getName()` but that probably requires braces.

ekrich commented on compiler/src/main/scala/org/scalajs/nscplugin/PrepJSExports.scala:
```diff
@@ -56,17 +56,19 @@ trait PrepJSExports[G <: Global with Singleton] { this: PrepJSInterop[G] =>
    *
-   *  * Registers top-level and static exports.
-   *  * Returns (non-static) exporters for this symbol.
+   *  * Registers top-level and static exports. * Returns (non-static) exporters
```
These appear to need to be dashes.

ekrich replied:
```scala
*  - Registers top-level and static exports.
*  - Returns (non-static) exporters for this symbol.
```

kitbellew replied:
sure, i can fix this

kitbellew commented on compiler/src/main/scala/org/scalajs/nscplugin/JSPrimitives.scala:
```diff
@@ -42,36 +43,36 @@ abstract class JSPrimitives {
-  final val JS_THROW = FORIN + 1                          // js.special.throw
-  final val JS_TRY_CATCH = JS_THROW + 1                   // js.special.tryCatch
-  final val WRAP_AS_THROWABLE = JS_TRY_CATCH + 1          // js.special.wrapAsThrowable
+  final val DYNAMIC_IMPORT = WITH_CONTEXTUAL_JS_CLASS_VALUE + 1 // runtime.dynamicImport
```
no mention of this needing alignment in the coding style guide

kitbellew commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -3346,9 +3344,8 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
-        implicit pos: Position): js.Tree = {
+        implicit pos: Position): js.Tree =
       genIsAsInstanceOf(genExpr(obj), obj.tpe, targs.head.tpe, cast)
-    }
```
i have removed this part of the rule (RedundantBraces). i will try to add more ranges to inserting all-blocks braces but it also requires changing some formatting rules as well, so it will take time.

kitbellew commented:
- rebased
- updated scalafmt draft version, with more fixes
- removed some rules as requested
- put manual formatting as two early commits, including character data patching change

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -646,19 +646,19 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
-          withInline(shouldMarkInline).
-          withNoinline(sym.hasAnnotation(NoinlineAnnotationClass))
+        withInline(shouldMarkInline).
+        withNoinline(sym.hasAnnotation(NoinlineAnnotationClass))
```
Not very important, but perhaps a manual adjustment to
```scala
      val optimizerHints = OptimizerHints.empty
        .withInline(shouldMarkInline)
        .withNoinline(sym.hasAnnotation(NoinlineAnnotationClass))
```
would work?

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -3522,7 +3531,7 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
           actualArg match {
             case js.VarRef(`formalArgName`) =>
-              // This is trivial assignment, we don't need it
+            // This is trivial assignment, we don't need it
```
```suggestion
              () // This is trivial assignment, we don't need it
```

sjrd commented on javalib/src/main/scala/java/net/URI.scala:
```diff
@@ -400,28 +401,28 @@ object URI {
-    ")(?:%[0-9a-z]+)?"
+    "(?:"   +
+      lelem + "{7}"     + block + "|" + // 1:2:3:4:5:6:7:8
+      lelem + "{1,7}:|" + // 1::                                        1:2:3:4:5:6:7::
```
Hum, the local config override didn't quite work here. Should it be instead:
```scala
// scalafmt: { align.tokens."+" = [{ code = "//" }] }
```
(without `{ code = "+" }`)?

sjrd commented on linker/jvm/src/main/scala/org/scalajs/linker/backend/closure/ClosureAstTransformer.scala:
```diff
@@ -463,7 +464,7 @@ private class ClosureAstTransformer(featureSet: FeatureSet,
       case tree =>
         transformBlock(List(tree), pos)
-    } (pos)
+    }(pos)
```
Is there a scalafmt config that would preserve it? Otherwise no big deal. We'll live with that style.

kitbellew replied:
nope. i could add it but what would it be called, and when would it apply?

sjrd replied:
I guess it would be `spaces.betweenCurlyAndParenArgs`. If `true`, then insert a space between `}(` and between `){` when they are the end and beginning of curried argument lists. I wouldn't be surprised if scalafmt already inserts the space in the `){` situation. It's very common in Scala, and it's a common style to insert a space there. This rule is the converse. Although I guess the common case can be explained more easily by the rule "always a space before `{` (except right after `(`?)".

But like I said, it's no big deal. We only have a few dozens of occurrences, almost all in `OptimizerCore.scala`.

kitbellew replied:
i took a look. basically, it's easy to do incorrectly but hard if correctly.

the nuance is interaction with the braces rewrite rule which turns one-liner `{ ... }` args into `(...)`, which then requires a potentially different whitespace choice.

the choice of space or no space is determined by the opening delimiter only, not what comes before it. which is why there's a bit of inconsistency in the logic, because for `) {` the brace determines it (and it's always a space today) and for `} (` the paren does (and for that there's a parameter).

if we had to consider the preceding delimiter as well, which could potentially change when the result is written out, this might lead to non-idempotent formatting. unless, of course, we instead decide we should always use a space between two consecutive arguments, whether or not they use parens or braces. which is not what you want.

sjrd commented on sbt-plugin/src/sbt-test/linker/anonfunction-compat/scala3-old-compiler-new-lib/Main.scala:
I suggest we exclude `sbt-plugin/src/sbt-test/` for now. Too many things are written in the wrong way, and scalafmt is doing the wrong thing with our wrong code.

sjrd reviewed (changes requested):
Reviewed up to `test-adapter/` excluded.

My biggest fear (`linker/`) is behind us :laughing:

sjrd commented on test-suite/js/src/test/require-scala2/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTestScala2.scala:
```diff
@@ -81,9 +81,9 @@ object NonNativeJSTypeTestScala2 {
-    } with Parent(parentParam2 = { sideEffects += foo.toString(); foo + bar })() {
+          val xyz = foo + bar
+          val yz = { sideEffects += xyz.toString(); xyz + 2 }
+        } with Parent(parentParam2 = { sideEffects += foo.toString(); foo + bar })() {
```
Oops, looks like scalafmt doesn't handle Scala 2 early initializers.

I guess we'll have to live with it, since I don't think anyone wants to implement that support now.

sjrd commented on test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/PromiseMock.scala:
```diff
@@ -207,40 +207,40 @@ object PromiseMock {
-              innerReject(e)
+        {
+          (innerResolve: js.Function1[B | Thenable[B], _],
+              innerReject: js.Function1[scala.Any, _]) =>
```
Looks like a bug? Not a blocker.

kitbellew commented on library/src/main/scala/scala/scalajs/js/ThisFunction.scala:
```diff
@@ -30,51 +30,204 @@ trait ThisFunction extends js.Function
 object ThisFunction {
   // scalastyle:off line.size.limit
 
-  implicit def fromFunction1[T1, R](f: scala.Function1[T1, R]): js.ThisFunction0[T1, R] = (x1: T1) => f(x1)
```
done

kitbellew commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/DataInputStreamTest.scala:
```diff
@@ -261,15 +261,16 @@ trait DataInputStreamTest {
       private var i: Int = 0
-      def read(): Int = if (i < length) { i += 1; i } else -1
+      def read(): Int = if (i < length) { i += 1; i }
+      else -1
```
done

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:
```diff
@@ -1018,7 +1020,7 @@ private[emitter] object CoreJSLib {
         } :::
 
-        defineFunction1(VarField.doubleToLong)(x => Return {
+        defineFunction1(VarField.doubleToLong){ x => Return {
```
Let's revert this change since we're not formatting this file anymore.

kitbellew replied:
ugh :) i made sure the file is not changed in any other commits. didn't see there were more changes in the same "manual" commit.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/unstable/IRContainerImpl.scala:
```diff
@@ -27,22 +27,22 @@ import org.scalajs.linker.interface.{IRContainer, IRFile}
    *  version).
    */
-  val path: String,
+    val path: String,
```
Looks like we need to manually adjust this for the indentation of the comment be right.

kitbellew reviewed (commented):
- added back a very conservative docstrings rewrite (indentation mostly)
- added back a very conservative remove-braces rewrite
- implemented logic to insert braces any multi-line expressions except for two-liner if-else in assign
- adjusted manual formatting and fixed `format: off`
- also, added `format: off` in the **right** ScalaJsPlugin (apparently, there are two)

gzm0 commented:
:shipit: :shipit: :shipit:

kitbellew commented:
- in yml, skip scalafmt for java 8; let's see what happens with windows...

kitbellew commented:
- old complicated scripts removed, put `cs launch` into one them
- also called `cs launch` directly in YML (there's a coursier action above, i am assuming it already installed what it had to)
- moved YML modifying commit after initial formatting

sjrd commented:
> are you saying that in yml, add `coursier/setup-action@v1` and then run `cs launch -java-opt ...`? and get rid of the shell script altogether? or put `cs launch` in the shell script?

Get rid of the shell scripts altogether. So no changes to `.gitignore` in the first commit. And no additions in `./scripts/` in the CI commit.

It seems you figured out the rest of what I meant ;)

sjrd commented:
At home, people can choose to install `scalafmt` however they see fit. For example, with coursier all I had to do was `cs install scalafmt` and now I can do `scalafmt --test` and that's all. No need for every repo out there to have their own platform-dependent scripts.
